### PR TITLE
multidag v1

### PR DESCRIPTION
**Pull request by yairsaban**

### CODE REVIEW

File: src/baseTransaction.ts [hunk 1]
```diff
@@ -15,6 +15,8 @@ export enum BaseTransactionName {
   NETWORK_FEE = 'NFBT',
   ROLLING_RESERVE = 'RRBT',
   RECEIVER = 'RBT',
+  GENERATE_TOKEN = 'TGBT',
```
Comment (YuvalCoti) [inline src/baseTransaction.ts]: Names:
Generate token => token generate fee
mint token fee => token mint fee

File: src/baseTransaction.ts [hunk 2]
```diff
@@ -35,17 +37,21 @@ type BaseTransactionTime = 'createTime';
 export interface BaseTransactionData {
   hash: string;
   addressHash: string;
+  currencyHash?: string;
   amount: string;
   createTime: number;
   name: BaseTransactionName;
   items?: Item[];
   encryptedMerchantName?: string;
+  serviceData?: any;
```
Comment (YuvalCoti) [inline src/baseTransaction.ts]: Add type

File: src/baseTransaction.ts [hunk 3]
```diff
@@ -77,19 +85,31 @@ export class BaseTransaction {
...
     this.addressHash = addressHash;
-    this.amount = amount.stripTrailingZeros();
-    this.createTime = utils.utcNowToSeconds();
     this.name = name;
+    this.currencyHash = currencyHash;
+    this.originalCurrencyHash = originalCurrencyHash;
+    this.serviceData = serviceData;
+    this.signerHash = signerHash;
+    this.amount = amount.stripTrailingZeros();
+    this.createTime = createTime || utils.utcNowToSeconds();
```
Comment (YuvalCoti) [inline src/baseTransaction.ts:112]: Check if can remove default value create time

File: src/signature.ts [hunk 4]
```diff
@@ -96,3 +127,188 @@ export class ClaimReward extends Signature {
     return utils.numberToByteArray(this.creationTime, 8);
   }
 }
+
+export class OriginatorSignature extends Signature {
+  private currencyName: string;
+  private currencySymbol: string;
+  private description: string;
+  private totalSupply: number;
```
Comment (YuvalCoti) [inline src/signature.ts:141]: normalize to int

File: src/signature.ts [hunk 5]
```diff
@@ -96,3 +127,188 @@ export class ClaimReward extends Signature {
...
+
+    this.currencyHash = currencyHash;
+    this.mintingAmount = mintingAmount;
+    this.instantTime = instantTime;
+    this.signingType = SigningType.MINTING_QUOTE;
+  }
+
+  public getBytes() {
+    const currencyHashBytes = utils.hexToBytes(this.currencyHash);
+    const mintingAmountBytes = utils.getBytesFromString(this.mintingAmount.toString());
```
Comment (YuvalCoti) [inline src/signature.ts:234]: verify that is an integer

File: src/signature.ts [hunk 6]
```diff
@@ -96,3 +127,188 @@ export class ClaimReward extends Signature {
...
+    const instantTimeBytes = utils.numberToByteArray(this.instantTime, 8)
+    const byteArraysToMerge = [currencyHashBytes, mintingAmountBytes, instantTimeBytes];
+
+    return utils.concatByteArrays(byteArraysToMerge);
+  }
+}
+
+export class MintQuoteDataSignature extends Signature {
+  private currencyHash: string;
+  private mintingAmount: number;
```
Comment (YuvalCoti) [inline src/signature.ts:244]: verify that is an interger

File: src/signature.ts [hunk 7]
```diff
@@ -96,3 +127,188 @@ export class ClaimReward extends Signature {
...
+  private feeAmount: number;
```
Comment (YuvalCoti) [inline src/signature.ts:245]: remove trailing zeros

File: src/signature.ts [hunk 8]
```diff
@@ -96,3 +127,188 @@ export class ClaimReward extends Signature {
...
+    this.feeAmount = feeAmount;
+    this.receiverAddress = receiverAddress;
+    this.instantTime = instantTime;
+    this.signingType = SigningType.MINTING_DATA;
+  }
+
+  public getBytes() {
+    const currencyHashBytes = utils.hexToBytes(this.currencyHash);
+    const mintingAmountBytes = utils.getBytesFromString(this.mintingAmount.toString());
+    const feeAmountBytes = utils.getBytesFromString(this.feeAmount.toString());
```
Comment (YuvalCoti) [inline src/signature.ts]: check if need to format as bigdecimal first and if there is a difference

File: src/transaction.ts [hunk 9]
```diff
@@ -16,6 +16,8 @@ export enum TransactionType {
   TRANSFER = 'Transfer',
   ZEROSPEND = 'ZeroSpend',
   CHARGEBACK = 'Chargeback',
+  TOKENGENERATION = 'TokenGeneration',
```
Comment (YuvalCoti) [inline src/transaction.ts]: add underscore between words

File: src/transaction.ts [hunk 10]
```diff
@@ -91,15 +94,16 @@ export class Transaction {
     transactionDescription = 'No description',
     userHash: string,
     type?: TransactionType,
-    createHash = true
+    createHash = true,
+    createTime?: number
```
Comment (YuvalCoti) [inline src/transaction.ts:123]: check if can remove create time

File: src/utils/financeUtils.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
+import axios from 'axios';
```
Comment (YuvalCoti) [inline src/utils/financeUtils.ts:1]: edit name to financial node utils

File: src/utils/financeUtils.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+import { BaseTransactionData } from '..';
+import { CotiError } from '../cotiError';
+import { Network } from './utils';
+
+export namespace financeUtils {
+    export async function getTokenGenerationFeeBT(tokenGenerationRequest: any, financeNodeUrlUrl: string){
```
Comment (YuvalCoti) [inline src/utils/financeUtils.ts]: remove one url

File: src/utils/financeUtils.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        try {
+            const { data } = await axios.post(`${financeNodeUrlUrl}/admin/token/generate`, tokenGenerationRequest, { headers });
+    
+            return new BaseTransactionData(data.tokenGenerationFee);
+        } catch (error) {
+          throw getErrorMessage(error);
+        }
+    }
+
+    export async function getTokenMintQuoteFee(tokenMintQuoteFeeRequest: any, financeNodeUrl: string) {
```
Comment (YuvalCoti) [inline src/utils/financeUtils.ts]: set type

File: src/utils/financeUtils.ts [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        try {
+            const { data } = await axios.post(`${financeNodeUrl}/admin/token/quote`, tokenMintQuoteFeeRequest);
+      
+            return data.mintingFeeQuote;
+          } catch (error) {
+            throw getErrorMessage(error);
+          }
+    }
+    
+    export async function getTokenMintFee(tokenMintFeeRequest: any, financeNodeUrl: string){
```
Comment (YuvalCoti) [inline src/utils/financeUtils.ts]: set type

File: src/utils/nodeUtils.ts [hunk 15]
```diff
@@ -250,12 +264,80 @@ export namespace nodeUtils {
       const response = await axios.put(`${api || nodeUrl[network].api}/exchange/userType`, updateUserTypeMessage, { headers });
       return response.data.message;
     } catch (error) {
-      const errorMessage = error.response && error.response.data ? error.response.data.errorMessage : error.message;
-      throw new NodeError(errorMessage, { debugMessage: `Error update user type at trust score node` });
+      throw getErrorMessage(error, 'Error update user type at trust score node', 'errorMessage');
+    }
+  }
+
+  export async function getUserTokenCurrencies(userHash: string, seed: string, api?: string, network: Network = 'mainnet') {
```
Comment (YuvalCoti) [inline src/utils/nodeUtils.ts]: check if can send without both api and network

File: src/utils/nodeUtils.ts [hunk 16]
```diff
@@ -250,12 +264,80 @@ export namespace nodeUtils {
...
+    } catch (error) {
+      throw getErrorMessage(error, 'Error get user token currencies', 'errorMessage');
     }
   }
 
-  function getErrorMessage(error: any) {
-    return error.response && error.response.data ? error.response.data.message : error.message;
+  export async function getTokenDetails(currencyHash: string, userHash: string, seed: string, api?: string, network: Network = 'mainnet') {
+    const tokenCurrencies = new TokenDetailsSignature(userHash, currencyHash);
+    const indexedWallet = new Wallet({ seed });
```
Comment (YuvalCoti) [inline src/utils/nodeUtils.ts]: allow user to initiate wallet with private key

File: src/utils/nodeUtils.ts [hunk 17]
```diff
@@ -250,12 +264,80 @@ export namespace nodeUtils {
...
+    try {
+      const { data } = await axios.post(`${api || nodeUrl[network].api}/currencies/token/details`, payload, { headers });
+      
+      return data;
+    } catch (error) {
+      throw getErrorMessage(error, 'Error get user token details', 'errorMessage');
+    }
+  }
+
+  export async function getTokenDetailsBySymbol(currencySymbol: string, userHash: string, seed: string, api?: string, network: Network = 'mainnet') {
```
Comment (YuvalCoti) [inline src/utils/nodeUtils.ts]: replace seed by wallet in all functions

File: src/utils/nodeUtils.ts [hunk 18]
```diff
@@ -250,12 +264,80 @@ export namespace nodeUtils {
...
+    const tokenCurrencies = new TokenDetailsSignature(userHash, undefined, currencySymbol);
```
Comment (YuvalCoti) [inline src/utils/nodeUtils.ts]: fix it to two signatures or an object of configuration

File: src/utils/tokenUtils.ts [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+        symbol: currencySymbol,
+        description: description,
+        totalSupply: totalSupply,
+        scale: scale,
+        originatorHash: userHash,
+        originatorSignature: signatureData
+      },
+      currencyTypeData: {
+        currencyType: currencyType,
+        createTime: Math.round(instantTime),
```
Comment (YuvalCoti) [inline src/utils/tokenUtils.ts]: remove round

File: src/utils/tokenUtils.ts [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+      userHash,
+      signature: signatureData
+    };
+
+    return mintQuoteRequest;
+  }
+
+  export async function getTokenMintFeeRequest(currencyHash: string, mintingAmount: number, feeAmount: number, walletAddressRecieveToken: string, userHash: string, seed: string): Promise<any> {
+    const instantTime = Math.floor(new Date().getTime() / 1000);
+    const instantTime2 = instantTime * 1000;
```
Comment (YuvalCoti) [inline src/utils/tokenUtils.ts]: decide on a better name instantTime 2

File: src/utils/transactionUtils.ts [hunk 21]
```diff
@@ -121,21 +127,45 @@ async function getFees<T extends IndexedAddress>(
...
-  let { addressBalance, addressPreBalance } = balanceObject[address];
-  const balance = new BigDecimal(`${addressBalance}`);
-  const preBalance = new BigDecimal(`${addressPreBalance}`);
+function addInputBaseTranction(balanceObject: any, address: string, amount: number, baseTransactions: BaseTransaction[], currencyHash?: string, tokensBalanceObject?: any, feeAddress?: string) {
+  let balance;
+  let preBalance;
+  let addressBalance;
+  let addressPreBalance;
+
+  if (currencyHash && address === feeAddress) {
```
Comment (YuvalCoti) [inline src/utils/transactionUtils.ts]: get currency hash of native and set it

File: src/utils/transactionUtils.ts [hunk 22]
```diff
@@ -181,5 +215,30 @@ async function addTrustScoreToTransaction<T extends IndexedAddress>(
     network,
     trustScoreNode
   );
+
   transaction.addTrustScoreMessageToTransaction(transactionTrustScoreData);
 }
+
+export async function transactionTokenGeneration(params: {
```
Comment (YuvalCoti) [inline src/utils/transactionUtils.ts:297]: remove if not necesery

**Review by YuvalCoti — CHANGES_REQUESTED** [review]

File: src/baseTransaction.ts (hunk 3 above)
Comment (yairsaban) [reply src/baseTransaction.ts:112]: cannot remove it since the creation time of the transaction and IBT needs to be the same =/

**Pull request merged**